[Security] [RememberMe] Add support for parallel requests doing remember-me re-authentication

Seldaek opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | yes
| New feature?  | yes ish <!-- please update src/**/CHANGELOG.md files -->
| Deprecations? | no <!-- please update UPGRADE-*.md and src/**/CHANGELOG.md files -->
| Tickets       | Fix #40971, Fix #28314, Fix #18384
| License       | MIT
| Doc PR        | symfony/symfony-docs#... <!-- required for new features -->

This is a possible implementation to gather feedback mostly..

`TokenVerifierInterface` naming is kinda bad perhaps.. But my goal would be to merge it in TokenProviderInterface for 6.0 so it's not so important. Not sure if/how to best indicate this in terms of deprecation notices.

Anyway wondering if this would be an acceptable implementation (ideally in an application I would probably override the new methods from DoctrineTokenProvider to something like this which is less of a hack and does expiration properly:

```php
    public function verifyToken(PersistentTokenInterface $token, string $tokenValue)
    {
        if (hash_equals($token->getTokenValue(), $tokenValue)) {
            return true;
        }

        if (!$this->cache->hasItem('rememberme-' . $token->getSeries())) {
            return false;
        }

        /** @var CacheItem $item */
        $item = $this->cache->getItem('rememberme-' . $token->getSeries());
        $oldToken = $item->get();
            
        return hash_equals($oldToken, $tokenValue);
    }

    public function updateExistingToken(PersistentTokenInterface $token, string $tokenValue, \DateTimeInterface $lastUsed): void
    {
        $this->updateToken($token->getSeries(), $tokenValue, $lastUsed);

        /** @var CacheItem $item */
        $item = $this->cache->getItem('rememberme-'.$token->getSeries());
        $item->set($token->getTokenValue());
        $item->expiresAfter(60);
        $this->cache->save($item);
    }
```

If you think it'd be fine to require optionally the cache inside DoctrineTokenProvider to enable this feature instead of the hackish way I did it, that'd be ok for me too.

The current `DoctrineTokenProvider` implementation of `TokenVerifierInterface` relies on the lucky fact that series are generated using `base64_encode(random_bytes(64))` which always ends in the `==` padding of base64, so that allowed me to store an alternative token value temporarily by replacing `==` with `_`.

Alternative implementation options:

1. Inject cache in `DoctrineTokenProvider` and do a proper implementation (as shown above) that way
2. Do not implement at all in `DoctrineTokenProvider` and let users who care implement this themselves.
3. Implement as a new `token_verifier` option that could be configured on the `firewall->remember_me` key so you can pass an implementation if needed, and possibly ship a default one using cache that could be autoconfigured
4. Add events that allow modifying the token to be verified, and allow receiving the newly updated token incl series, instead of TokenVerifierInterface, but then we need to inject a dispatcher in RememberMeAuthenticator.

@chalasr @wouterj sorry for the long description but in the hope of getting this included in 5.3.0, if you can provide guidance I will happily work on this further tomorrow to try and wrap it up ASAP.

heiglandreas commented on src/Symfony/Bridge/Doctrine/Security/RememberMe/DoctrineTokenProvider.php:
```diff
@@ -136,6 +137,56 @@ public function createNewToken(PersistentTokenInterface $token)
+            return false;
+        }
+
+        if ($tmpToken->getLastUsed()->getTimestamp() + 60 < time()) {
```
Perhaps a coment why the `60` here? That we assume that the old token is invalid after 60 seconds, so for a whole minute it would be possible to use the old **and** the new `rememberMe`-token. 

If that is a critical issue we somehow need to make that configurable.

heiglandreas reviewed (approved):
Looks good to me. Only the comment regarding the hardcoded 60 seconds is something that could be considered. But that is not a reason to not merge this from my point of view.

derrabus commented on src/Symfony/Component/Security/Core/Authentication/RememberMe/TokenVerifierInterface.php:
```diff
@@ -0,0 +1,34 @@
+     * This lets you override the token check logic to for example accept slightly outdated tokens.
+     *
+     * Do not forget to implement token comparisons using hash_equals for a secure implementation.
+     *
+     * @return bool
+     */
+    public function verifyToken(PersistentTokenInterface $token, string $tokenValue);
```
```suggestion
     */
    public function verifyToken(PersistentTokenInterface $token, string $tokenValue): bool;
```

derrabus commented on src/Symfony/Bridge/Doctrine/Security/RememberMe/DoctrineTokenProvider.php:
```diff
@@ -136,6 +137,56 @@ public function createNewToken(PersistentTokenInterface $token)
+    /**
+     * {@inheritDoc}
+     */
+    public function verifyToken(PersistentTokenInterface $token, string $tokenValue)
```
```suggestion
    public function verifyToken(PersistentTokenInterface $token, string $tokenValue): bool
```

derrabus reviewed (changes requested):
Please make use of return types for brand-new methods. Also, can you please update the component's changelog?

Seldaek commented:
I updated the implementation to make `token_verifier` configurable in the security config (definitely would appreciate a review of that bit from @wouterj). By default if the token_provider implements TokenVerifierInterface it will also be used as token verifier, so that makes the default implementation in DoctrineTokenProvider work by default. 

I could remove the DoctrineTokenProvider implementation though if you rather avoid having a default fix which is slightly hackish (I fully documented the implementation now so it's hopefully clearer what is going on).

I also added a CacheTokenVerifier implementation which is the ideal one I would configure in prod, but I guess this can not be enabled by default as it requires a configured cache? I'm not sure what the baseline is here.

Edit: I'm also not sure why Psalm fails, it complains about some code I did not change, not sure if I should apply those fixes or not.

wouterj commented:
fyi: I don't have time for a detailed review today, it's on my list for tomorrow

wouterj commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/RememberMeFactory.php:
```diff
@@ -214,6 +216,14 @@ public function addConfiguration(NodeDefinition $node)
                     ->end()
                 ->end()
+            ->end()
+            ->arrayNode('token_verifier')
+                ->beforeNormalization()
+                    ->ifString()->then(function ($v) { return ['service' => $v]; })
+                ->end()
+                ->children()
+                    ->scalarNode('service')->info('The service ID of a custom rememberme token verifier.')->end()
+                ->end()
             ->end();
 
```
I would prefer to make this a `token_verifier` scalar node. Making it an array later on (if we want to add more options) is easy to make BC - using the technique you show here.

Seldaek replied:
Absolutely makes sense, this was a copy paste from token_provider above sorry :)

wouterj commented on src/Symfony/Component/Security/Core/Authentication/RememberMe/CacheTokenVerifier.php:
```diff
@@ -0,0 +1,71 @@
+ */
+class CacheTokenVerifier implements TokenVerifierInterface
+{
+    /** @var CacheItemPoolInterface */
+    private $cache;
+    /** @var int */
+    private $outdatedTokenTtl;
+    /** @var string */
+    private $cacheKeyPrefix;
```
Please remove the PHPdoc from these properties.

In Symfony, we always skip the PHPdocs if they don't introduce anything that isn't already documented by the constructor (through typehints or PHPdoc)

wouterj commented on src/Symfony/Component/Security/Core/Authentication/RememberMe/CacheTokenVerifier.php:
```diff
@@ -0,0 +1,71 @@
+    private $cacheKeyPrefix;
+
+    /**
+     * @param int $outdatedTokenTtl How long should the outdated token be valid, defaults to 60
+     *                              which matches how often the PersistentRememberMeHandler will
+     *                              at most refresh tokens. Increasing to more than that is not
+     *                              recommended, but you may use a lower value.
```
```suggestion
     * @param int $outdatedTokenTtl How long the outdated token should still be considered valid. Defaults
     *                              to 60, which matches how often the PersistentRememberMeHandler will at
     *                              most refresh tokens. Increasing to more than that is not recommended,
     *                              but you may use a lower value.
```

wouterj commented on src/Symfony/Component/Security/Http/RememberMe/PersistentRememberMeHandler.php:
```diff
@@ -66,7 +73,11 @@ public function processRememberMe(RememberMeDetails $rememberMeDetails, UserInte
-        if (!hash_equals($persistentToken->getTokenValue(), $tokenValue)) {
+
+        if (
+            ($this->tokenVerifier && !$this->tokenVerifier->verifyToken($persistentToken, $tokenValue))
+            || !hash_equals($persistentToken->getTokenValue(), $tokenValue)
```
Wouldn't it be better to swap the OR condition, given that `verifyToken` probably involves calling a db/cache and so adds more load on the server than the `hash_equals`?

Seldaek replied:
Oh sorry this was a mistake, I meant to only call the verifier if it is set, so that it is fully in control of token verification, and hash_equals is only there as fallback if no verifier is present. I dumbed the code down a bit now so it's clearer and does what I meant it to :)

Note that any sane implementation (and the two included) would call hash_equals first and if the token is valid (which it is in 99.9% of cases) it returns early so there is really no performance impact.

wouterj commented on src/Symfony/Component/Security/Http/RememberMe/PersistentRememberMeHandler.php:
```diff
@@ -32,13 +33,19 @@
 final class PersistentRememberMeHandler extends AbstractRememberMeHandler
 {
     private $tokenProvider;
+    /** @var ?TokenVerifierInterface */
```
```suggestion
```

Seldaek commented:
Thanks for the review @wouterj - it's all taken care of. 

Still would be curious to get an opinion on these:

> I could remove the DoctrineTokenProvider implementation though if you rather avoid having a default fix which is slightly hackish (I fully documented the implementation now so it's hopefully clearer what is going on).

> I also added a CacheTokenVerifier implementation which is the ideal one I would configure in prod, but I guess this can not be enabled by default as it requires a configured cache? I'm not sure what the baseline is here.

My ideal outcome would be to enable the CacheTokenVerifier by default and remove the other one as I find it rather suboptimal, but not sure if it's possible.

chalasr commented on src/Symfony/Component/Security/Core/Authentication/RememberMe/CacheTokenVerifier.php:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    /**
+     * {@inheritDoc}
```
```suggestion
     * {@inheritdoc}
```
per our coding standards

chalasr commented on src/Symfony/Component/Security/Core/Authentication/RememberMe/CacheTokenVerifier.php:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    /**
+     * {@inheritDoc}
```
```suggestion
     * {@inheritdoc}
```

chalasr commented on src/Symfony/Bridge/Doctrine/Security/RememberMe/DoctrineTokenProvider.php:
```diff
@@ -136,6 +137,65 @@ public function createNewToken(PersistentTokenInterface $token)
     }
 
+    /**
+     * {@inheritDoc}
```
```suggestion
     * {@inheritdoc}
```

chalasr commented on src/Symfony/Component/Security/Core/Authentication/RememberMe/CacheTokenVerifier.php:
```diff
@@ -0,0 +1,68 @@
+/**
+ * @author Jordi Boggiano <j.boggiano@seld.be>
+ */
+class CacheTokenVerifier implements TokenVerifierInterface
```
We probably need a dedicated "system" cache pool to wire this one as a service.
Here is an example we could take inspiration from: https://github.com/symfony/symfony/blob/1a87c72c1f67e909a59c8dbe5ba3cce71a5bbeea/src/Symfony/Bundle/SecurityBundle/Resources/config/security.php#L278-L282
And the conditional removal of the service for the case where cache is not available:
https://github.com/symfony/symfony/blob/fd02ec075d68d54ff69e948d021223234d70e367/src/Symfony/Bundle/SecurityBundle/DependencyInjection/Compiler/AddExpressionLanguageProvidersPass.php#L37-L39

chalasr commented:
> My ideal outcome would be to enable the CacheTokenVerifier by default and remove the other one as I find it rather suboptimal, but not sure if it's possible

Let's keep it for the case where the cache integration is not enabled, which should be pretty rare.

wouterj commented:
Sorry, I forgot my global review message.

I like the `token_verifier` option and providing the optimal implementation in the code. That seems to be a good middle-ground for this not-so-common-but-complex issue.

> My ideal outcome would be to enable the CacheTokenVerifier by default and remove the other one as I find it rather suboptimal, but not sure if it's possible.

Yeah, I'm afraid we don't want to force the dependency on cache. I'm still not very sure if this is a must-fix issue, or rather an edge-case. Would not having a token verifier and only shipping the CacheTokenVerifier make sense?

Seldaek commented:
Thanks for the hints @chalasr - I hope my implementation in 8c1d0ccf56 is correct - I'm not entirely sure if the nested null on invalid ref will work here https://github.com/symfony/symfony/pull/41175/files#diff-c790b1d460cd1f019b3eefa9c0322710d078dbadd510dbb1ee631c4c668e0dbcR323-R328 I'm hoping it would set the cache ref to null if the cache service gets deleted by the pass, which would then mark the whole verifier service invalid as the cache arg is not nullable, turning the whole verifier into a null itself.

@wouterj I think we can argue either way.. but if the current implementation works I'd say why not ship a fix by default and save people from having to find out about this issue before they can configure the verifier service.

chalasr commented:
>  I'm not entirely sure if the nested null on invalid ref will work here https://github.com/symfony/symfony/pull/41175/files#diff-c790b1d460cd1f019b3eefa9c0322710d078dbadd510dbb1ee631c4c668e0dbcR323-R328 I'm hoping it would set the cache ref to null if the cache service gets deleted by the pass, which would then mark the whole verifier service 

@Seldaek `NULL_ON_INVALID_REFERENCE` means that if the referenced service does not exist, null will be injected as argument to the referencing service.
If we want to remove the CacheTokenVerifier service (I think so), we need to call `removeDefinition()` from the compiler pass you added.

chalasr commented:
You will probably need a DI tag to collect the cache-based token verifier definitions and remove them from the compiler pass.

Seldaek commented:
Yes but injecting null for $cache means the CacheTokenVerifier definition becomes invalid, so I believe it will automatically be removed/nulled, and the RememberMeAuthenticator receives null as verifier.

I tested with this and it seems to work as expected, I get an `App\Test2` instance with null arg:

```yaml
services:
    App\Test: # this would be the verifier service
        arguments:
            - '@?cache.foo'

    App\Test2: # and this the authenticator
        arguments:
            - '@?App\Test'
```

chalasr commented:
In your example `App\Test2` should either be given a `App\Test` instance or an exception should be thrown if the arg is not nullable (which is the case here), but `App\Test` won't be removed:
```
App\Test2 {#1367 ▼
  -test: App\Test {#6687 ▼
    -cacheItemPool: null
  }
```
Here is a reproducer https://github.com/chalasr/repro-null-on-invalid (clone, run `symfony serve` and browse localhost:8000). Do I miss something?

Seldaek commented:
Nope, App\Test2 accepts a nullable App\Test in my example, because that's what the PersistentRememberMeHandler accepts https://github.com/symfony/symfony/pull/41175/files#diff-5128596229172c9c8dc0dda0e9846231169b9d239c65366cb13ce6ffc2a913daR39 (a nullable TokenVerifierInterface). Anyway if you want I can tweak the compiler pass to explicitly nullify the service, but to me it does not seem necessary.

Edit: You got the example backwards  I now realize, App\Test should get a non-nullable cache pool, because that's also what CacheTokenVerifier expects.

chalasr commented:
Ok got it, thanks!

chalasr commented on src/Symfony/Bundle/SecurityBundle/DependencyInjection/Security/Factory/RememberMeFactory.php:
```diff
@@ -304,4 +311,20 @@ private function createTokenProvider(ContainerBuilder $container, string $firewa
         return $tokenProviderId;
     }
+
+    private function createTokenVerifier(ContainerBuilder $container, string $firewallName, ?string $serviceId): string
```
```suggestion
    private function createTokenVerifier(ContainerBuilder $container, string $firewallName, ?string $serviceId): Reference
```
I'm currently trying your patch, looks good apart from that string cast 👍

Seldaek replied:
Woops, thanks for spotting that.

nicolas-grekas commented:
That's for 5.4, or 5.3?

Seldaek commented:
I was hoping for 5.3 but that's just my opinion ;)

zerkms commented:
So it won't be backported to 4.4?

wouterj commented:
@zerkms no, Symfony only fixes bugs in already released versions.

I'm a bit divided on 5.3/5.4. On one hand, it's a nice and mostly hidden "bug fix" that fits nicely with the other remember me changes in 5.3. On the other hand, the change is fully BC and thus has no reason (other than another 6 months waiting) to bypass the feature-freeze deadline (unless I'm missing something).

zerkms commented:
@wouterj it says "3 year support for bugs and security fixes." at https://symfony.com/releases for 4.4 :shrug:

chalasr commented:
@zerkms It won't be backported because the patch is not compatible with the 4.4 implementation of the remember-me feature.

chalasr reviewed (approved):
This is fine-tuning the 5.3 rewrite of the remember-me feature, while removing a long standing limitation of that feature, so I think it's fine for 5.3.

zerkms commented:
@chalasr then

> 3 year support for bugs and security fixes.

should be reworded, to be like "3 year support for _some_ bugs and security fixes"?

If a bug is reported against v4.4 (which also is affected) would it be fixed independently?

chalasr commented:
If someone is able to submit a valid fix for 4.4, we could probably accept it.
But this seems to be design issue that is hard to fix on the 4.4 implementation.

fabpot commented:
Tests seem to be broken.

fabpot commented:
Thank you @Seldaek.
